api/alarms/index.js

```diff
@@ -0,0 +1,182 @@
+
+// DOM event bindings
+
+//// Alarm display buttons
+
+clearButton.addEventListener('click', () => manager.cancelAllAlarms());
+refreshButton.addEventListener('click', () => manager.refreshDisplay());
+
+//// New alarm form
+
+form.addEventListener('submit', (event) => {
+  event.preventDefault();
+  let formData = new FormData(form);
+  let data = Object.fromEntries(formData);
+
+  // Extract form values
+  let name = data['alarm-name'];
+  let delay = Number.parseFloat(data['time-value']);
+  let delayFormat = data['time-format'];
+  let period = Number.parseFloat(data['period']);
+
+  // Prepare alarm info for creation call
+  let alarmInfo = {};
+
+  if (delayFormat === 'ms') {
+    // Specified in milliseconds, use `when` property
+    alarmInfo.when = Date.now() + delay;
+  } else if (delayFormat === 'min') {
+    // specified in minutes, use `delayInMinutes` property
+    alarmInfo.delayInMinutes = delay;
+  }
+
+  if (period) {
+    alarmInfo.periodInMinutes = period;
+  }
+
+  // Create the alarm – this uses the same signature as chrome.alarms.create
+  manager.createAlarm(name, alarmInfo);
+});
+
+class AlarmManager {
+  constructor({display, log} = {}) {
+    if (display) {
+      this.displayElement = display;
+    }
+    if (log) {
+      this.logElement = log;
+    }
+
+    this.logMessage('Manager: initializing demo');
+
+    this.displayElement.addEventListener('click', (event) => this.handleCancelAlarm(event));
+    chrome.alarms.onAlarm.addListener(alarm => this.handleAlarm(alarm));
+  }
+
+  logMessage(message) {
+    let date = new Date();
+    let pad = (val, len = 2) => val.toString().padStart(len, '0');
+    let h = pad(date.getHours());
+    let m = pad(date.getMinutes());
+    let s = pad(date.getSeconds());
+    let ms = pad(date.getMilliseconds(), 3);
+    let time = `${h}:${m}:${s}.${ms}`;
+
+    let logLine = document.createElement('div');
+    logLine.innerText = `[${time}] ${message}`;
+
+    // Log events in reverse chronological order
+    this.logElement.insertBefore(logLine, this.logElement.firstChild);
+
+    // Log events in chronological order
+    // this.logElement.appendChild(logLine);
+  }
+
+  async handleAlarm(alarm) {
+    let json = JSON.stringify(alarm, null, 2).replace(/\s+/g, ' ');
+    this.logMessage(`Alarm "${alarm.name}" fired\n${json}}`);
+    await this.refreshDisplay();
+  }
+
+  async handleCancelAlarm(event) {
+    if (!event.target.classList.contains('alarm-row__cancel-button')) {
+      return;
+    }
+
+    let name = event.target.parentElement.dataset.name;
+    await this.cancelAlarm(name);
+    await this.refreshDisplay();
+  }
+
+  async cancelAlarm(name) {
+    // TODO: Remove custom promise wrapper once the Alarms API supports promises
+    return new Promise((resolve, reject) => {
+      chrome.alarms.clear(name, (wasCleared) => {
+        if (wasCleared) {
+          this.logMessage(`Manager: canceled alarm "${name}"`)
```

Comment:
nit: missing trailing semicolon here and below

Reply:
Fixed.